File: flutter/packages/workout_log_core/lib/src/analytics/session_metrics.dart
import '../coding.dart';
import '../models/block.dart';
import '../models/session.dart';
import '../models/work_set.dart';
import '../wall_clock.dart';

/// Which energy system a metcon taxed, by how long it lasted (P9).
///
/// Classified on total duration, which is all the file records. An interval
/// workout of short efforts therefore lands in the domain its *clock* says, not
/// the one its work intervals say — those live in the block's notes.
enum TimeDomain {
  alactic, // <= 5 min
  glycolytic, // 5-20 min
  aerobic; // > 20 min

  static TimeDomain? forSeconds(double? seconds) => switch (seconds) {
        null => null,
        <= 0 => null,
        <= 300 => TimeDomain.alactic,
        <= 1200 => TimeDomain.glycolytic,
        _ => TimeDomain.aerobic,
      };
}

/// Work time against clock time, from the bracket timestamps (P1).
class SessionDensity {
  const SessionDensity({
    required this.workMin,
    required this.transitionMin,
    this.totalMin,
  });

  final double workMin;
  final double transitionMin;

  /// Null when the session has no start time or nothing that ends: density is
  /// derived from what was written down, never guessed.
  final double? totalMin;

  double? get workShare {
    final total = totalMin;
    if (total == null || total <= 0) return null;
    return _round(workMin / total, 3);
  }

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{
      'work_min': _round(workMin),
      'transition_min': _round(transitionMin),
    };
    put(json, 'total_min', _round(totalMin));
    put(json, 'work_share', workShare);
    return json;
  }
}

class BlockMetrics {
  const BlockMetrics({
    required this.index,
    required this.type,
    this.label,
    this.durationMin,
    this.transitionMin,
    this.sets = 0,
    this.backoffSets = 0,
    this.reps = 0,
    this.tonnageKg = 0,
    this.band,
  });

  final int index;
  final String type;

  /// The exercise, the machine, or the metcon's format — whatever names the block.
  final String? label;
  final double? durationMin;
  final double? transitionMin;
  final int sets;
  final int backoffSets;
  final int reps;
  final double tonnageKg;

  /// The slot's own rep band (P6): the band most of its sets sit in.
  final RepBand? band;

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{'index': index, 'type': type};
    put(json, 'label', label);
    put(json, 'duration_min', _round(durationMin));
    put(json, 'transition_min', _round(transitionMin));
    if (sets > 0) json['sets'] = sets;
    if (backoffSets > 0) json['backoff_sets'] = backoffSets;
    if (reps > 0) json['reps'] = reps;
    if (tonnageKg > 0) json['tonnage_kg'] = _round(tonnageKg);
    put(json, 'rep_band', band?.name);
    return json;
  }
}

/// One round of a metcon with its split differenced out of the cumulative
/// clock the paper log records.
class RoundSplit {
  const RoundSplit({
    required this.round,
    this.reps,
    this.cumulativeSec,
    this.splitSec,
    this.heartRate,
  });

  final int round;
  final int? reps;
  final double? cumulativeSec;
  final double? splitSec;
  final int? heartRate;

  double? get secondsPerRep {
    final split = splitSec;
    final count = reps;
    if (split == null || count == null || count <= 0) return null;
    return _round(split / count, 2);
  }

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{'round': round};
    put(json, 'reps', reps);
    put(json, 'cumulative_sec', _round(cumulativeSec));
    put(json, 'split_sec', _round(splitSec));
    put(json, 'sec_per_rep', secondsPerRep);
    put(json, 'heart_rate', heartRate);
    return json;
  }
}

class MetconMetrics {
  MetconMetrics({
    required this.blockIndex,
    this.format,
    this.scheme,
    this.exercises = const [],
    this.totalSec,
    this.rounds = const [],
  });

  final int blockIndex;
  final String? format;
  final List<int>? scheme;
  final List<String> exercises;
  final double? totalSec;
  final List<RoundSplit> rounds;

  TimeDomain? get domain => TimeDomain.forSeconds(totalSec);

  /// How much slower the last round ran than the first, per rep, as a percentage
  /// — the pacing question a 21-15-9 is asked (positive = faded).
  double? get paceDecayPercent {
    final paces = [
      for (final round in rounds)
        if (round.secondsPerRep != null) round.secondsPerRep!,
    ];
    if (paces.length < 2 || paces.first <= 0) return null;
    return _round((paces.last - paces.first) / paces.first * 100, 1);
  }

  int? get peakHeartRate {
    final rates = [
      for (final round in rounds)
        if (round.heartRate != null) round.heartRate!,
    ];
    if (rates.isEmpty) return null;
    return rates.reduce((a, b) => a > b ? a : b);
  }

  Map<String, dynamic> toJson() {
    final json = <String, dynamic>{'block_index': blockIndex};
    put(json, 'format', format);
    put(json, 'scheme', scheme);
    if (exercises.isNotEmpty) json['exercises'] = exercises;
    put(json, 'total_sec', _round(totalSec));
    put(json, 'time_domain', domain?.name);
    put(json, 'pace_decay_percent', paceDecayPercent);
    put(json, 'peak_heart_rate', peakHeartRate);
    if (rounds.isNotEmpty) {
      json['rounds'] = [for (final r in rounds) r.toJson()];
    }
    return json;
  }
}

/// Everything one session says about itself once the derivations are run:
/// tonnage, rep bands, density and metcon splits. Nothing here is stored — the
/// file holds only what was observed (ADR-001).
class SessionMetrics {
  SessionMetrics({
    required this.date,
    required this.cycleDay,
    required this.kind,
    required this.blocks,
    required this.metcons,
    required this.density,
    required this.tonnageKg,
    required this.sets,
    required this.backoffSets,
    required this.reps,
    required this.bandSets,
    required this.bandSlots,
  });

  factory SessionMetrics.of(Session session) {
    final blocks = <BlockMetrics>[];
    final metcons = <MetconMetrics>[];
    final bandSets = <RepBand, int>{};
    final bandSlots = <RepBand, int>{};
    var tonnage = 0.0;
    var sets = 0;
    var backoffSets = 0;
    var reps = 0;
    var workMin = 0.0;
    var transitionMin = 0.0;

    final sessionStart = WallClock.minutes(session.startTime);
    int? cursor = sessionStart;
    int? lastEnd;

    for (var index = 0; index < session.blocks.length; index++) {
      final block = session.blocks[index];
      final start = WallClock.minutes(_startTimeOf(block)) ?? cursor;
      final end = WallClock.minutes(_endTimeOf(block));

      double? duration;
      if (start != null && end != null && end >= start) {
        duration = (end - start).toDouble();
      } else if (block is CardioBlock) {
        duration = block.durationMin;
      }

      double? transition;
      if (cursor != null && start != null && start > cursor) {
        transition = (start - cursor).toDouble();
      }

      if (duration != null) workMin += duration;
      if (transition != null) transitionMin += transition;

      if (end != null) {
        cursor = end;
        lastEnd = end;
      } else if (start != null && duration != null) {
        cursor = start + duration.round();
        lastEnd = cursor;
      }

      switch (block) {
        case StrengthBlock(:final exercise, sets: final blockSets):
          final counts = <RepBand, int>{};
          var blockReps = 0;
          var blockTonnage = 0.0;
          var blockBackoff = 0;
          for (final set in blockSets) {
            final band = set.band;
            if (band != null) {
              counts[band] = (counts[band] ?? 0) + 1;
              bandSets[band] = (bandSets[band] ?? 0) + 1;
            }
            blockReps += set.recordedReps ?? 0;
            blockTonnage += set.tonnageKg;
            if (set.isBackoff ?? false) blockBackoff++;
          }
          final slotBand = _dominantBand(counts);
          if (slotBand != null) {
            bandSlots[slotBand] = (bandSlots[slotBand] ?? 0) + 1;
          }
          sets += blockSets.length;
          backoffSets += blockBackoff;
          reps += blockReps;
          tonnage += blockTonnage;
          blocks.add(BlockMetrics(
            index: index,
            type: block.type,
            label: exercise,
            durationMin: duration,
            transitionMin: transition,
            sets: blockSets.length,
            backoffSets: blockBackoff,
            reps: blockReps,
            tonnageKg: blockTonnage,
            band: slotBand,
          ));
        case MetconBlock():
          final metcon = _metconMetrics(index, block, start, end);
          metcons.add(metcon);
          blocks.add(BlockMetrics(
            index: index,
            type: block.type,
            label: block.format?.wire ?? 'metcon',
            durationMin: duration ??
                (metcon.totalSec == null ? null : metcon.totalSec! / 60),
            transitionMin: transition,
          ));
        case CardioBlock(:final machine):
          blocks.add(BlockMetrics(
            index: index,
            type: block.type,
            label: machine.isEmpty ? null : machine,
            durationMin: duration,
            transitionMin: transition,
          ));
        case CooldownBlock():
          blocks.add(BlockMetrics(
            index: index,
            type: block.type,
            durationMin: duration,
            transitionMin: transition,
          ));
      }
    }

    final totalMin = (sessionStart != null && lastEnd != null)
        ? (lastEnd - sessionStart).toDouble()
        : null;

    return SessionMetrics(
      date: session.date,
      cycleDay: session.cycleDay,
      kind: session.kind,
      blocks: blocks,
      metcons: metcons,
      density: SessionDensity(
        workMin: workMin,
        transitionMin: transitionMin,
        totalMin: totalMin,
      ),
      tonnageKg: tonnage,
      sets: sets,
      backoffSets: backoffSets,
      reps: reps,
      bandSets: bandSets,
      bandSlots: bandSlots,
    );
  }

  final String date;
  final String cycleDay;
  final Kind kind;
  final List<BlockMetrics> blocks;
  final List<MetconMetrics> metcons;
  final SessionDensity density;
  final double tonnageKg;
  final int sets;
  final int backoffSets;
  final int reps;

  /// Sets per band, and slots (strength blocks) per band. Both matter: P6 tracks
  /// sets, P4's "2-3 volume slots per cycle" counts slots.
  final Map<RepBand, int> bandSets;
  final Map<RepBand, int> bandSlots;

  /// The ramp's top set is the progress signal, back-offs are volume (P2), so
  /// they are never averaged together.
  int get topSets => sets - backoffSets;

  Map<String, dynamic> toJson() => {
        'date': date,
        'cycle_day': cycleDay,
        'kind': kind.name,
        'strength': {
          'tonnage_kg': _round(tonnageKg),
          'sets': sets,
          'top_sets': topSets,
          'backoff_sets': backoffSets,
          'reps': reps,
          'rep_band_sets': _bandJson(bandSets),
          'rep_band_slots': _bandJson(bandSlots),
        },
        'density': density.toJson(),
        'blocks': [for (final b in blocks) b.toJson()],
        if (metcons.isNotEmpty)
          'metcons': [for (final m in metcons) m.toJson()],
      };
}

Map<String, int> _bandJson(Map<RepBand, int> counts) => {
      for (final band in RepBand.values)
        if ((counts[band] ?? 0) > 0) band.name: counts[band]!,
    };

/// Ties go to the heavier band: a slot that is half sixes and half eights is
/// not evidence for the volume dose P4 rations.
RepBand? _dominantBand(Map<RepBand, int> counts) {
  RepBand? best;
  var bestCount = 0;
  for (final band in RepBand.values) {
    final count = counts[band] ?? 0;
    if (count > bestCount) {
      best = band;
      bestCount = count;
    }
  }
  return best;
}

MetconMetrics _metconMetrics(
  int index,
  MetconBlock block,
  int? startMin,
  int? endMin,
) {
  final rounds = <RoundSplit>[];
  double? previous;
  for (final round in block.rounds ?? const <MetconRound>[]) {
    final cumulative = round.splitCumulativeSec;
    double? split;
    if (cumulative != null) {
      split = previous == null ? cumulative : cumulative - previous;
      previous = cumulative;
    }
    rounds.add(RoundSplit(
      round: round.round,
      reps: round.reps ?? _schemeReps(block, round.round),
      cumulativeSec: cumulative,
      splitSec: split,
      heartRate: round.heartRate,
    ));
  }

  final fromRounds = rounds.isEmpty ? null : rounds.last.cumulativeSec;
  final fromClock = (startMin != null && endMin != null && endMin >= startMin)
      ? (endMin - startMin) * 60.0
      : null;

  return MetconMetrics(
    blockIndex: index,
    format: block.format?.wire,
    scheme: block.scheme,
    exercises: [for (final e in block.exercises) e.name],
    totalSec: fromRounds ?? fromClock,
    rounds: rounds,
  );
}

/// The reps a round carries when the round itself does not say: the scheme
/// position it matches (21-15-9 → round 2 is 15).
int? _schemeReps(MetconBlock block, int round) {
  final scheme = block.scheme;
  if (scheme == null || round < 1 || round > scheme.length) return null;
  return scheme[round - 1];
}

String? _startTimeOf(Block block) => switch (block) {
      StrengthBlock(:final startTime) => startTime,
      MetconBlock(:final startTime) => startTime,
      CardioBlock() || CooldownBlock() => null,
    };

String? _endTimeOf(Block block) => switch (block) {
      CardioBlock(:final endTime) => endTime,
      StrengthBlock(:final endTime) => endTime,
      MetconBlock(:final endTime) => endTime,
      CooldownBlock(:final endTime) => endTime,
    };

double? _round(double? value, [int digits = 1]) {
  if (value == null) return null;
  final factor = [1, 10, 100, 1000][digits];
  return (value * factor).round() / factor;
}

File: crates/workout-log-core/src/analytics/session_metrics.rs
//! Everything one session says about itself once the derivations are run:
//! tonnage, rep bands, density and metcon splits. Nothing here is stored — the
//! file holds only what was observed (ADR-001).

use serde_json::{Map, Value, json};

use crate::models::block::Block;
use crate::models::metcon::MetconBlock;
use crate::models::session::{Kind, Session};
use crate::models::work_set::RepBand;
use crate::wall_clock;

/// Which energy system a metcon taxed, by how long it lasted (P9).
///
/// Classified on total duration, which is all the file records. An interval
/// workout of short efforts therefore lands in the domain its *clock* says.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TimeDomain {
    Alactic,
    Glycolytic,
    Aerobic,
}

impl TimeDomain {
    pub const ALL: [Self; 3] = [Self::Alactic, Self::Glycolytic, Self::Aerobic];

    pub fn name(self) -> &'static str {
        match self {
            Self::Alactic => "alactic",
            Self::Glycolytic => "glycolytic",
            Self::Aerobic => "aerobic",
        }
    }

    pub fn for_seconds(seconds: Option<f64>) -> Option<Self> {
        match seconds {
            Some(s) if s <= 0.0 => None,
            Some(s) if s <= 300.0 => Some(Self::Alactic),
            Some(s) if s <= 1200.0 => Some(Self::Glycolytic),
            Some(_) => Some(Self::Aerobic),
            None => None,
        }
    }
}

fn round(value: f64, digits: u32) -> f64 {
    let factor = 10f64.powi(digits as i32);
    (value * factor).round() / factor
}

fn round_opt(value: Option<f64>, digits: u32) -> Option<f64> {
    value.map(|v| round(v, digits))
}

fn put(json: &mut Map<String, Value>, key: &str, value: Option<Value>) {
    if let Some(value) = value {
        json.insert(key.to_string(), value);
    }
}

/// Work time against clock time, from the bracket timestamps (P1).
#[derive(Debug, Clone, PartialEq)]
pub struct SessionDensity {
    pub work_min: f64,
    pub transition_min: f64,
    /// None when the session has no start time or nothing that ends: density is
    /// derived from what was written down, never guessed.
    pub total_min: Option<f64>,
}

impl SessionDensity {
    pub fn work_share(&self) -> Option<f64> {
        match self.total_min {
            Some(total) if total > 0.0 => Some(round(self.work_min / total, 3)),
            _ => None,
        }
    }

    pub fn to_json(&self) -> Value {
        let mut json = Map::new();
        json.insert("work_min".into(), json!(round(self.work_min, 1)));
        json.insert(
            "transition_min".into(),
            json!(round(self.transition_min, 1)),
        );
        put(
            &mut json,
            "total_min",
            round_opt(self.total_min, 1).map(|v| json!(v)),
        );
        put(&mut json, "work_share", self.work_share().map(|v| json!(v)));
        Value::Object(json)
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct BlockMetrics {
    pub index: usize,
    pub block_type: String,
    /// The exercise, the machine, or the metcon's format — whatever names the block.
    pub label: Option<String>,
    pub duration_min: Option<f64>,
    pub transition_min: Option<f64>,
    pub sets: usize,
    pub backoff_sets: usize,
    pub reps: i64,
    pub tonnage_kg: f64,
    /// The slot's own rep band (P6): the band most of its sets sit in.
    pub band: Option<RepBand>,
}

impl BlockMetrics {
    pub fn to_json(&self) -> Value {
        let mut json = Map::new();
        json.insert("index".into(), json!(self.index));
        json.insert("type".into(), json!(self.block_type));
        put(&mut json, "label", self.label.clone().map(|v| json!(v)));
        put(
            &mut json,
            "duration_min",
            round_opt(self.duration_min, 1).map(|v| json!(v)),
        );
        put(
            &mut json,
            "transition_min",
            round_opt(self.transition_min, 1).map(|v| json!(v)),
        );
        if self.sets > 0 {
            json.insert("sets".into(), json!(self.sets));
        }
        if self.backoff_sets > 0 {
            json.insert("backoff_sets".into(), json!(self.backoff_sets));
        }
        if self.reps > 0 {
            json.insert("reps".into(), json!(self.reps));
        }
        if self.tonnage_kg > 0.0 {
            json.insert("tonnage_kg".into(), json!(round(self.tonnage_kg, 1)));
        }
        put(
            &mut json,
            "rep_band",
            self.band.map(|b| json!(band_name(b))),
        );
        Value::Object(json)
    }
}

/// One round of a metcon with its split differenced out of the cumulative
/// clock the paper log records.
#[derive(Debug, Clone, PartialEq)]
pub struct RoundSplit {
    pub round: i64,
    pub reps: Option<i64>,
    pub cumulative_sec: Option<f64>,
    pub split_sec: Option<f64>,
    pub heart_rate: Option<i64>,
}

impl RoundSplit {
    pub fn seconds_per_rep(&self) -> Option<f64> {
        let split = self.split_sec?;
        let count = self.reps?;
        (count > 0).then(|| round(split / count as f64, 2))
    }

    pub fn to_json(&self) -> Value {
        let mut json = Map::new();
        json.insert("round".into(), json!(self.round));
        put(&mut json, "reps", self.reps.map(|v| json!(v)));
        put(
            &mut json,
            "cumulative_sec",
            round_opt(self.cumulative_sec, 1).map(|v| json!(v)),
        );
        put(
            &mut json,
            "split_sec",
            round_opt(self.split_sec, 1).map(|v| json!(v)),
        );
        put(
            &mut json,
            "sec_per_rep",
            self.seconds_per_rep().map(|v| json!(v)),
        );
        put(&mut json, "heart_rate", self.heart_rate.map(|v| json!(v)));
        Value::Object(json)
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct MetconMetrics {
    pub block_index: usize,
    pub format: Option<String>,
    pub scheme: Option<Vec<i64>>,
    pub exercises: Vec<String>,
    pub total_sec: Option<f64>,
    pub rounds: Vec<RoundSplit>,
}

impl MetconMetrics {
    pub fn domain(&self) -> Option<TimeDomain> {
        TimeDomain::for_seconds(self.total_sec)
    }

    /// How much slower the last round ran than the first, per rep, as a
    /// percentage — the pacing question a 21-15-9 is asked (positive = faded).
    pub fn pace_decay_percent(&self) -> Option<f64> {
        let paces: Vec<f64> = self
            .rounds
            .iter()
            .filter_map(RoundSplit::seconds_per_rep)
            .collect();
        let (first, last) = (*paces.first()?, *paces.last()?);
        (paces.len() >= 2 && first > 0.0).then(|| round((last - first) / first * 100.0, 1))
    }

    pub fn peak_heart_rate(&self) -> Option<i64> {
        self.rounds.iter().filter_map(|r| r.heart_rate).max()
    }

    pub fn to_json(&self) -> Value {
        let mut json = Map::new();
        json.insert("block_index".into(), json!(self.block_index));
        put(&mut json, "format", self.format.clone().map(|v| json!(v)));
        put(&mut json, "scheme", self.scheme.clone().map(|v| json!(v)));
        if !self.exercises.is_empty() {
            json.insert("exercises".into(), json!(self.exercises));
        }
        put(
            &mut json,
            "total_sec",
            round_opt(self.total_sec, 1).map(|v| json!(v)),
        );
        put(
            &mut json,
            "time_domain",
            self.domain().map(|d| json!(d.name())),
        );
        put(
            &mut json,
            "pace_decay_percent",
            self.pace_decay_percent().map(|v| json!(v)),
        );
        put(
            &mut json,
            "peak_heart_rate",
            self.peak_heart_rate().map(|v| json!(v)),
        );
        if !self.rounds.is_empty() {
            json.insert(
                "rounds".into(),
                Value::Array(self.rounds.iter().map(RoundSplit::to_json).collect()),
            );
        }
        Value::Object(json)
    }
}

pub type BandCounts = [usize; 3];

fn band_index(band: RepBand) -> usize {
    match band {
        RepBand::Heavy => 0,
        RepBand::Base => 1,
        RepBand::Volume => 2,
    }
}

fn band_name(band: RepBand) -> &'static str {
    match band {
        RepBand::Heavy => "heavy",
        RepBand::Base => "base",
        RepBand::Volume => "volume",
    }
}

const BAND_ORDER: [RepBand; 3] = [RepBand::Heavy, RepBand::Base, RepBand::Volume];

fn band_json(counts: &BandCounts) -> Value {
    let mut json = Map::new();
    for band in BAND_ORDER {
        let count = counts[band_index(band)];
        if count > 0 {
            json.insert(band_name(band).into(), json!(count));
        }
    }
    Value::Object(json)
}

/// Ties go to the heavier band: a slot that is half sixes and half eights is
/// not evidence for the volume dose P4 rations.
fn dominant_band(counts: &BandCounts) -> Option<RepBand> {
    let mut best = None;
    let mut best_count = 0;
    for band in BAND_ORDER {
        let count = counts[band_index(band)];
        if count > best_count {
            best = Some(band);
            best_count = count;
        }
    }
    best
}

#[derive(Debug, Clone, PartialEq)]
pub struct SessionMetrics {
    pub date: String,
    pub cycle_day: String,
    pub kind: Kind,
    pub blocks: Vec<BlockMetrics>,
    pub metcons: Vec<MetconMetrics>,
    pub density: SessionDensity,
    pub tonnage_kg: f64,
    pub sets: usize,
    pub backoff_sets: usize,
    pub reps: i64,
    /// Sets per band, and slots (strength blocks) per band. Both matter: P6
    /// tracks sets, P4's "2-3 volume slots per cycle" counts slots.
    pub band_sets: BandCounts,
    pub band_slots: BandCounts,
}

impl SessionMetrics {
    /// The ramp's top set is the progress signal, back-offs are volume (P2), so
    /// they are never averaged together.
    pub fn top_sets(&self) -> usize {
        self.sets - self.backoff_sets
    }

    pub fn of(session: &Session) -> Self {
        let mut blocks = Vec::new();
        let mut metcons = Vec::new();
        let mut band_sets: BandCounts = [0; 3];
        let mut band_slots: BandCounts = [0; 3];
        let (mut tonnage, mut sets, mut backoff_sets, mut reps) = (0.0, 0usize, 0usize, 0i64);
        let (mut work_min, mut transition_min) = (0.0, 0.0);

        let session_start = wall_clock::minutes(session.start_time.as_deref());
        let mut cursor = session_start;
        let mut last_end: Option<i64> = None;

        for (index, block) in session.blocks.iter().enumerate() {
            let start = wall_clock::minutes(block.start_time()).or(cursor);
            let end = wall_clock::minutes(block.end_time());

            let duration = match (start, end) {
                (Some(s), Some(e)) if e >= s => Some((e - s) as f64),
                _ => match block {
                    Block::Cardio(cardio) => cardio.duration_min,
                    _ => None,
                },
            };
            let transition = match (cursor, start) {
                (Some(c), Some(s)) if s > c => Some((s - c) as f64),
                _ => None,
            };

            if let Some(d) = duration {
                work_min += d;
            }
            if let Some(t) = transition {
                transition_min += t;
            }

            if let Some(e) = end {
                cursor = Some(e);
                last_end = Some(e);
            } else if let (Some(s), Some(d)) = (start, duration) {
                cursor = Some(s + d.round() as i64);
                last_end = cursor;
            }

            match block {
                Block::Strength(strength) => {
                    let mut counts: BandCounts = [0; 3];
                    let (mut block_reps, mut block_tonnage, mut block_backoff) =
                        (0i64, 0.0, 0usize);
                    for set in &strength.sets {
                        if let Some(band) = set.band() {
                            counts[band_index(band)] += 1;
                            band_sets[band_index(band)] += 1;
                        }
                        block_reps += set.recorded_reps().unwrap_or(0);
                        block_tonnage += set.tonnage_kg();
                        if set.is_backoff.unwrap_or(false) {
                            block_backoff += 1;
                        }
                    }
                    let slot_band = dominant_band(&counts);
                    if let Some(band) = slot_band {
                        band_slots[band_index(band)] += 1;
                    }
                    sets += strength.sets.len();
                    backoff_sets += block_backoff;
                    reps += block_reps;
                    tonnage += block_tonnage;
                    blocks.push(BlockMetrics {
                        index,
                        block_type: "strength".into(),
                        label: Some(strength.exercise.clone()),
                        duration_min: duration,
                        transition_min: transition,
                        sets: strength.sets.len(),
                        backoff_sets: block_backoff,
                        reps: block_reps,
                        tonnage_kg: block_tonnage,
                        band: slot_band,
                    });
                }
                Block::Metcon(metcon) => {
                    let derived = metcon_metrics(index, metcon, start, end);
                    let duration_min = duration.or(derived.total_sec.map(|s| s / 60.0));
                    blocks.push(BlockMetrics {
                        index,
                        block_type: "metcon".into(),
                        label: Some(
                            metcon
                                .format
                                .map_or("metcon".to_string(), |f| format_wire(f).to_string()),
                        ),
                        duration_min,
                        transition_min: transition,
                        ..empty_block(index, "metcon")
                    });
                    metcons.push(derived);
                }
                Block::Cardio(cardio) => blocks.push(BlockMetrics {
                    label: (!cardio.machine.is_empty()).then(|| cardio.machine.clone()),
                    duration_min: duration,
                    transition_min: transition,
                    ..empty_block(index, "cardio")
                }),
                Block::Cooldown(_) => blocks.push(BlockMetrics {
                    duration_min: duration,
                    transition_min: transition,
                    ..empty_block(index, "cooldown")
                }),
            }
        }

        let total_min = match (session_start, last_end) {
            (Some(start), Some(end)) => Some((end - start) as f64),
            _ => None,
        };

        Self {
            date: session.date.clone(),
            cycle_day: session.cycle_day.clone(),
            kind: session.kind,
            blocks,
            metcons,
            density: SessionDensity {
                work_min,
                transition_min,
                total_min,
            },
            tonnage_kg: tonnage,
            sets,
            backoff_sets,
            reps,
            band_sets,
            band_slots,
        }
    }

    pub fn to_json(&self) -> Value {
        let mut json = Map::new();
        json.insert("date".into(), json!(self.date));
        json.insert("cycle_day".into(), json!(self.cycle_day));
        json.insert("kind".into(), json!(kind_name(self.kind)));
        json.insert(
            "strength".into(),
            json!({
                "tonnage_kg": round(self.tonnage_kg, 1),
                "sets": self.sets,
                "top_sets": self.top_sets(),
                "backoff_sets": self.backoff_sets,
                "reps": self.reps,
                "rep_band_sets": band_json(&self.band_sets),
                "rep_band_slots": band_json(&self.band_slots),
            }),
        );
        json.insert("density".into(), self.density.to_json());
        json.insert(
            "blocks".into(),
            Value::Array(self.blocks.iter().map(BlockMetrics::to_json).collect()),
        );
        if !self.metcons.is_empty() {
            json.insert(
                "metcons".into(),
                Value::Array(self.metcons.iter().map(MetconMetrics::to_json).collect()),
            );
        }
        Value::Object(json)
    }
}

fn empty_block(index: usize, block_type: &str) -> BlockMetrics {
    BlockMetrics {
        index,
        block_type: block_type.into(),
        label: None,
        duration_min: None,
        transition_min: None,
        sets: 0,
        backoff_sets: 0,
        reps: 0,
        tonnage_kg: 0.0,
        band: None,
    }
}

pub fn kind_name(kind: Kind) -> &'static str {
    match kind {
        Kind::Training => "training",
        Kind::Deload => "deload",
        Kind::Retest => "retest",
    }
}

pub fn format_wire(format: crate::models::metcon::MetconFormat) -> &'static str {
    use crate::models::metcon::MetconFormat as F;
    match format {
        F::ForTime => "for_time",
        F::Amrap => "amrap",
        F::Emom => "emom",
        F::Intervals => "intervals",
        F::Ladder => "ladder",
        F::Chipper => "chipper",
    }
}

fn metcon_metrics(
    index: usize,
    block: &MetconBlock,
    start_min: Option<i64>,
    end_min: Option<i64>,
) -> MetconMetrics {
    let mut rounds = Vec::new();
    let mut previous: Option<f64> = None;
    for round in block.rounds.iter().flatten() {
        let cumulative = round.split_cumulative_sec;
        let split = cumulative.map(|c| previous.map_or(c, |p| c - p));
        if cumulative.is_some() {
            previous = cumulative;
        }
        rounds.push(RoundSplit {
            round: round.round,
            reps: round.reps.or_else(|| scheme_reps(block, round.round)),
            cumulative_sec: cumulative,
            split_sec: split,
            heart_rate: round.heart_rate,
        });
    }

    let from_rounds = rounds.last().and_then(|r| r.cumulative_sec);
    let from_clock = match (start_min, end_min) {
        (Some(s), Some(e)) if e >= s => Some((e - s) as f64 * 60.0),
        _ => None,
    };

    MetconMetrics {
        block_index: index,
        format: block.format.map(|f| format_wire(f).to_string()),
        scheme: block.scheme.clone(),
        exercises: block.exercises.iter().map(|e| e.name.clone()).collect(),
        total_sec: from_rounds.or(from_clock),
        rounds,
    }
}

/// The reps a round carries when the round itself does not say: the scheme
/// position it matches (21-15-9 → round 2 is 15).
fn scheme_reps(block: &MetconBlock, round: i64) -> Option<i64> {
    let scheme = block.scheme.as_ref()?;
    if round < 1 || round as usize > scheme.len() {
        return None;
    }
    Some(scheme[round as usize - 1])
}
